File: Day_1/BankAccountManagement.py
from typing import Dict

class BankAccountManagement:
    def __init__(self):
        print("Bank Account Management Server is Started")

    class BankAccount:
        def __init__(self,name,balance):
            self.createAccount(name,balance)
            
        def createAccount(self,name,balance):
            self.name=name
            self.balance = balance
        
        def deposit(self,amount):
            if amount < 0:
                raise ValueError("Deposit amount must be greater than 0")
            self.balance += amount
            print("Amount: " + str(amount) + "/- is Successfully Deposited to "+self.name)
            print("New Balance: " + str(self.balance) + "/-")
        
        def withdraw(self,amount) -> int:
            if amount < 0:
                raise ValueError("Withdrawal amount must be greater than 0")
            if amount > self.balance:
                raise ValueError("Insufficient Balance")
            self.balance-=amount
            print("Amount: " + str(amount) + "/- is Successfully Withdrawn from "+ self.name)
            print("New Balance: " + str(self.balance) + "/-")
            return amount
        
        def getBalance(self) -> int:
            return self.balance
        
        def updateName(self,name):
            self.name = name
        
        def getName(self) -> str:
            return self.name
        
        def getDetails(self)->Dict[str,str]:
            datas = dict()
            datas.update({"Name":self.name});
            datas.update({"Balance":str(self.balance)})
            return datas

if __name__=="__main__":
    karaikalBranch = BankAccountManagement()

    ac1 = karaikalBranch.BankAccount("Kedhara Murugavel V", 100000)
    ac2 = karaikalBranch.BankAccount("Kalaiselvi K",100000)
    ac3 = karaikalBranch.BankAccount("Niranjani K",100000)
    ac4 = karaikalBranch.BankAccount("Naleaswara Kumaran K",100000)

    ac4withdrawal = ac4.withdraw(75000)

    split3 = ac4withdrawal/3

    ac1.deposit(split3)
    ac2.deposit(split3)
    ac3.deposit(split3)

    acs = [ac1,ac2,ac3,ac4]

    for ac in acs:
        data = ac.getDetails()
        print(data)
    

File: Day_1/BankAccountManagement.dart
void main() {
  Bankaccountmanagement karaikalBranch = Bankaccountmanagement();
  BankAccount ac1 = karaikalBranch.createAccount("Kedhara Murugavel", 100000);
  BankAccount ac2 = karaikalBranch.createAccount("Kalaiselvi", 100000);
  BankAccount ac3 = karaikalBranch.createAccount("Niranjani", 100000);
  BankAccount ac4 = karaikalBranch.createAccount("Naleaswara Kumaran K", 100000);

  int ac4Withdrawal = ac4.withdraw(75000);
  int splitinto3 =  ac4Withdrawal~/3;
  ac1.deposit(splitinto3);
  ac2.deposit(splitinto3);
  ac3.deposit(splitinto3);

  List<BankAccount> acs = [ac1,ac2,ac3,ac4];

  acs.forEach((ac){
    print(ac.getDetails());
  });
}

class Bankaccountmanagement {
  BankAccountManagement(){
    print("Bank Account Management Server is Started");
  }
  BankAccount createAccount(String name, int balance){
    return BankAccount(name,balance);
  }
}

class BankAccount {
  late String name;
  late int balance;
  
  BankAccount(String name, int balance) {
    this.createAccount(name, balance);
  }

  void createAccount(String name, balance) {
    this.name = name;
    this.balance = balance;
  }

  void deposit(int amount) {
    if (amount < 0) {
      throw Exception("Amount should be greater than 0");
    }
    print("Amount: " + amount.toString() + "/- is Successfully Deposited to "+this.name);
    print("New Balance: " + this.balance.toString() + "/-");
    this.balance += amount;
  }

  int withdraw(int amount) {
    if (amount < 0) {
      throw Exception("Amount should be greater than 0");
    }
    if (amount > balance) {
      throw Exception("Insufficient Balance");
    }

    this.balance -= amount;
    print("Amount: " + amount.toString() + "/- is Successfully Withdrawn from "+this.name);
    print("New Balance: " + this.balance.toString() + "/-");
    return amount;
  }

  int getBalance() {
    return this.balance;
  }

  void updateName(String name) {
    this.name = name;
  }

  String getName() {
    return this.name;
  }

  Map<String, String> getDetails(){
    Map<String,String> datas = {};
    datas["Names"]=this.getName();
    datas["Balance"]=this.getBalance().toString();
    return datas;
  }
}
